File: src/ispeech/model/ModelSine.cc
/**
 * @file ModelSine.cpp
 *
 * A wrapper for the FFT algorithm found in Ooura mathematical packages.
 *
 * This file is part of the ispeech library.
 * ispeech is iqiyi software, licensed under the MIT/X11 License. A copy of
 * the license is provided with the library in the LICENSE file.
 *
 * @package vfx_ispeech
 * @version 0.0.1-dev
 * @author Han Zhongtao
 * @date 2019-2019
 * @license http://www.opensource.org/licenses/mit-license.php MIT
 * @since 0.0.1
 */

#include "ModelSine.h"
#include <algorithm>
#include <cmath>
#include <cstddef>
#include <random>

namespace vfx_ispeech
{
  /**
   * Initializes a model.
   *
   * Prepares for a DFT model.
   *
   * @param length input signal size (usually a power of 2)
   */
  ModelSine::ModelSine()
  {
  }

  /**
   * Destroys the model object and cleans up working area.
   */
  ModelSine::~ModelSine()
  {
  }

  /**
   * Returns a complex data spectrum value.
   *
   * @param x input signal
   * @param fs sampling rate
   * @param w analysis window
   * @param N size of complex spectrum,it means the FFT size too
   * @param t threshold in nagative dB
   * @return output array sound
   */
  SampleListType ModelSine::sineModel(SignalSource &x,double fs,SignalSource &w,std::size_t N,double t)
  {
     std::size_t window_size = w.size();

     //half analysis window size by rounding
     std::size_t hM1 = (std::size_t)std::floor((window_size+1)/2);//128

     //half analysis window size by floor
     std::size_t hM2 = (std::size_t)std::floor(window_size/2);//128

     //FFT size for synthesis(even)
     std::size_t Ns = 512;

     //Hop size used for analysis and synthesis
     std::size_t H = Ns/4;
     //half of synthesis FFT size
     std::size_t hNs = Ns/2;

     //init sound pointer in middle of anal window
     std::size_t pin = std::max(hNs,hM1);
     //last sample to start a frame
     std::size_t pend = x.size()-std::max(hNs,hM1);

     LOG_ERR(this,"ModelSine::sineModel pin = %lld,pend = %lld",pin,pend);
     //initialize buffer for FFT
     //SampleListType fftbuffer(N);
     //zeroSamples(fftbuffer);

     //initialize output sound frame
     SampleListType yw(Ns,0);
     //initialize output array
     SampleListType y(x.size(),0);

     //normalize analysis window
     w/=sum(w);

     //initialize analysis window
     SampleListType sw(Ns,0);

     //triangular window
     TriangWindow ow(2*H);

     //add triangular window
     for(std::size_t i = hNs-H,j=0;i<hNs+H;i++,j++){
       sw[i]=ow.sample(j);
     }

     //blackmanharris window
     BlackmanharrisWindow bh(Ns);

     //normalize blackmanharris window
     bh/=sum(bh);

     //normalized synthesis window
     for(std::size_t i = 0;i<Ns;i++){
       sw[i]/=bh.sample(i);
     }

     ModelDFT dft;
     OouraFft fft(Ns);

     while(pin < pend){
       //analysis
       SignalSource x1(x.toArray()+pin-hM1,window_size);
       AnalysisType a = dft.Analysis(x1,w,N);
       SampleListType ploc = peakDetection(a.mX,t);

       PeakInterpolateType ipeaks = peakInterp(a.mX,a.pX,ploc);

       /*
	 std::size_t ysize = ipeaks.iploc.size();
	 LOG_ERR(this,"ModelSine::sineModel ysize = %lld",ysize);

	 SampleListType out(ysize);
	 for(std::size_t i = 0; i < ysize;i++)
	 {
	 out[i]= ipeaks.iploc[i];
	 LOG_ERR(this,"ModelSine::sineModel out[%lld] = %f",i,out[i]);

	 }

	 return out;
       */

       SampleListType ipfreq(ipeaks.iploc.size());

       for(std::size_t i = 0 ; i < ipeaks.iploc.size(); i++){
	 ipfreq[i]=fs*ipeaks.iploc[i]/float(N);
       }
       //synthesis
       SpectrumType Y = genSpecSines(ipfreq,ipeaks.ipmag,ipeaks.ipphase,Ns,fs);

       /*
	 std::size_t ysize = Y.size();
	 LOG_ERR(this,"ModelSine::sineModel ysize = %lld",ysize);

	 SampleListType out(ysize);
	 for(std::size_t i = 0; i < ysize;i++)
	 {
	 out[i]= Y[i].real();
	 LOG_ERR(this,"ModelSine::sineModel out[%lld] = %e",i,out[i]);

	 }
       */

       SampleListType fftbuffer = fft.ifft2(Y);
       /*
	 std::size_t ysize = fftbuffer.size();
	 LOG_ERR(this,"ModelSine::sineModel ysize = %lld",ysize);

	 SampleListType out(ysize);
	 for(std::size_t i = 0; i < ysize;i++)
	 {
	 out[i]= fftbuffer[i];
	 LOG_ERR(this,"ModelSine::sineModel out[%lld] = %f",i,out[i]);

	 }
	 return out;
       */
       for(std::size_t i = hNs+1,fi=0;fi < hNs-1; i++,fi++)
	 {
	   yw[fi] = fftbuffer[i];
	 }

       for(std::size_t i = 0,fi=hNs-1;i < hNs; i++,fi++)
	 {
	   yw[fi] = fftbuffer[i];
	 }

       for(std::size_t i = pin-hNs,k=0;i<pin+hNs; i++,k++)
	 {
	   y[i]+=sw[k]*yw[k];
	 }

       pin+=H;
     }

    return y;
  }

  /**
   * Returns the mugnitude and phase of signal at frequency-domain.
   *
   * @param x input signal
   * @param w analysis window
   * @param N FFT size
   * @return AnalysisType mX is the mugnitude of spectrum,pX is the phase of spectrum.
   */
  AnalysisType ModelSine::Analysis(SignalSource &x,double fs,SignalSource &w,std::size_t N,std::size_t H,double t,std::size_t maxnSines /*= 100*/,double minSineDur/*= 0.01*/,double freqDevOffset /*= 20.0*/,double freqDevSlope/*=0.01*/)
  {
    xtfreq.clear();
    xtmag.clear();
    xtphase.clear();

    if(minSineDur < 0){
      AnalysisType e;
      return e;
    }

     std::size_t window_size = w.size();

     if(window_size > N){
       LOG_ERR(this,"ModelSine::Analysis Window size (M=%lld) is bigger than FFT size (N=%lld)!",window_size,N);

       AnalysisType e;
       return e;
     }

    //if all x == 0 return 0.

     std::size_t hN = N/2 + 1;//257
     std::size_t hM1 = (window_size + 1) / 2;//128
     std::size_t hM2 = (std::size_t)std::floor(window_size/2);//128

     SignalSource zeros(hM2);
     // add zeros at beginning to center first window at sample 0
     x.appendHead(zeros);
     // add zeros at the end to analyze last sample
     x.append(zeros);

     //init sound pointer in middle of anal window
     std::size_t pin = hM1;
     //last sample to start a frame
     std::size_t pend = x.size()- hM1;

     //LOG_ERR(this,"ModelSine::Analysis pin = %lld,pend = %lld",pin,pend);

     //normalize analysis window
     w/=sum(w);

     SampleListType tfreq;

     ModelDFT dft;

     while( pin < pend){
       // select frame
       SignalSource x1(x.toArray()+pin-hM1,window_size);
       // compute dft
       AnalysisType a = dft.Analysis(x1,w,N);
       // detect locations of peaks
       SampleListType ploc = peakDetection(a.mX,t);
       // refine peak values by interpolation
       PeakInterpolateType ipeaks = peakInterp(a.mX,a.pX,ploc);

       // convert peak locations to Hertz
       SampleListType ipfreq(ipeaks.iploc.size());
       for(std::size_t i = 0 ; i < ipeaks.iploc.size(); i++){
	 ipfreq[i]=fs*ipeaks.iploc[i]/(float)N;
       }

       //LOG_ERR(this,"ModelSine::Analysis after peakInterp the result :ipeaks.iploc.size  = %lld",ipeaks.iploc.size());

       // perform sinusoidal tracking by adding peaks to trajectories
       SineTrackingType  tracking  = sineTracking(ipfreq, ipeaks.ipmag, ipeaks.ipphase, tfreq, freqDevOffset, freqDevSlope);


       tfreq = tracking.tfreqn;
       if(tfreq.size() > maxnSines)
	 {
	   tfreq.erase(tfreq.begin()+std::min(maxnSines,tfreq.size()),tfreq.end());
	 }
       else
	 {
	   for(std::size_t k = tfreq.size(); k < maxnSines; k++)
	     {
	       tfreq.push_back(0.0);
	     }
	 }

       std::size_t ysize = tfreq.size();
       //LOG_ERR(nullptr,"Functions sineTracking finally tracked number = %lld",ysize);
       /*
       for(std::size_t k = 0; k < ysize;k++) {
	 LOG_ERR(nullptr,"Functions sineTracking finally tfreqn[%lld] = %f",k,tfreq[k]);
       }
       */

       SampleListType tfreqn(maxnSines,0.0);
       SampleListType tmagn(maxnSines,0.0);
       SampleListType tphasen(maxnSines,0.0);

       for(std::size_t k = 0; k < maxnSines; k++)
	 {
	   if(tfreq.size() > k)
	     {
	       tfreqn[k] = tfreq[k];
	     }
	   else
	     {
	       tfreqn[k] = 0.0;
	     }

	   if(tracking.tmagn.size() > k)
	     {
	       tmagn[k] = tracking.tmagn[k];
	     }
	   else
	     {
	       tmagn[k] = 0.0;
	     }

	   if(tracking.tphasen.size() > k)
	     {
	       tphasen[k] = tracking.tphasen[k];
	     }
	   else
	     {
	       tphasen[k] = 0.0;
	     }
	 }

       xtfreq.push_back(tfreqn);
       xtmag.push_back(tmagn);
       xtphase.push_back(tphasen);

       pin+=H;
     }

     //clean sine tracks shorter than minSineDur
     cleaningSineTracks(xtfreq,fs*minSineDur/H);

    AnalysisType result;

    return result;
  }

  std::size_t ModelSine::GetAnalysisTypeCount()
  {
    return xtfreq.size();
  }

  SineTrackingType ModelSine::GetAnalysisTypeByIndex(std::size_t index)
  {
    SineTrackingType  tracking;
    tracking.tfreqn = xtfreq[index];
    tracking.tmagn = xtmag[index];
    tracking.tphasen = xtphase[index];

    return tracking;
  }

  /**
   * Returns the synthesis signal.
   *
   * @param tfreq frequencies of sinusoids
   * @param tmag magnitudes of sinusoids
   * @param tphase phases of sinusoids
   * @param N Synthesis FFT size
   * @param H hop size
   * @param fs sampling rate
   * @return synthesis of a signal using the discrete Fourier transfom
   */
  SampleListType ModelSine::Synthesis(SampleList2DType &tfreq,SampleList2DType &tmag,SampleList2DType tphase,std::size_t N,std::size_t H,double fs)
  {
    //half of FFT size for synthesis
    std::size_t hN = N/2;

    // number of frames
    std::size_t L = tfreq.size();

    //initialize output sound pointer
    std::size_t pout = 0;
    //output sound size
    std::size_t ysize = H*(L+3);
    //initialize output array
    SampleListType y(ysize,0);
    LOG_ERR(nullptr,"ModelSine::Synthesis ysize = %lld",ysize);
    //initialize synthesis window
    SampleListType sw(N,0);

    //triangular window
    TriangWindow ow(2*H);

     //put triangular window at the center of sw
     for(std::size_t i = hN-H,j=0;i<hN+H;i++,j++){
       sw[i]=ow.sample(j);
     }

     //blackmanharris window
     BlackmanharrisWindow bh(N);

     //normalize blackmanharris window
     //because of generate signal we will use the main lob of blackmanharris,so had to use hb to inverse.
     bh/=sum(bh);

     //normalized synthesis window
     for(std::size_t i = 0;i < N;i++){
       sw[i]/=bh.sample(i);
     }

     //initialize synthesis frequencies
     SampleListType lastytfreq(tfreq[0].begin(),tfreq[0].end());
     //initialize synthesis phases
    std::random_device rd;
    std::map<int, int> hist;
    std::uniform_int_distribution<int> dist(0, 99);
    SampleListType ytphase(tfreq[0].size());

    for(std::size_t k = 0; k < tfreq[0].size(); k++)
      {
	//random float value between 0.0 and 2pi
	ytphase[k] = 2*M_PI*dist(rd)/100.0;
      }

    OouraFft fft(N);
    for(std::size_t l = 0; l < L; l++)
      {
	if(tphase.size() > 0)
	  {
	    for(std::size_t k = 0; k < tphase[l].size(); k++)
	      {
		ytphase[k]=tphase[l][k];
	      }
	  }
	else
	  {
	    for(std::size_t k = 0; k < tfreq[l].size(); k++)
	      {
		ytphase[k] +=(M_PI*(lastytfreq[k]+tfreq[l][k])/fs)*H;
	      }
	  }

	SpectrumType Y = genSpecSines(tfreq[l],tmag[l],ytphase,N,fs);

	for(std::size_t k = 0; k < tfreq[l].size(); k++)
	  {
	    lastytfreq[k]=tfreq[l][k];
	  }

	for(std::size_t k = 0; k < ytphase.size(); k++)
	  {
	    ytphase[k] = fmod(ytphase[k],2*M_PI);
	  }

	//ifft
	SampleListType fftbuffer = fft.ifft2(Y);

	//fftshift
	SampleListType yw(N);
	for(std::size_t fi = 0,i = hN+1; fi < hN-1; i++,fi++ )
	  {
	    yw[fi] = fftbuffer[i];
	  }
	for(std::size_t fi=hN-1,i=0; i < hN;i++,fi++)
	  {
	    yw[fi]=fftbuffer[i];
	  }

	//
	for(std::size_t k = 0; k < N; k++)
	  {
	    y[pout+k]+=sw[k]*yw[k];
	  }
	pout+=H;
      }

    y.erase(y.begin(),y.begin()+hN);
    y.erase(y.begin()+y.size()-hN,y.end());

    SampleListType result;
    for(size_t i = hN;i<y.size()-hN;i++)
      {
	//result.push_back(y[i]);
      }
    //return result;
    return y;
  }
  SampleListType ModelSine::AnalysisAndSynthesis(SignalSource &x,double fs,SignalSource &w,std::size_t N,std::size_t H,double t,std::size_t maxnSines /*= 100*/,double minSineDur/*= 0.01*/,double freqDevOffset /*= 20.0*/,double freqDevSlope/*=0.01*/)
  {
    Analysis(x,fs,w,N,H,t,maxnSines,minSineDur,freqDevOffset,freqDevSlope);
    LOG_ERR(nullptr,"ModelSine::AnalysisAndSynthesis Analysis done");

    N=512;

    SampleListType y = Synthesis(xtfreq,xtmag,xtphase,N,H,fs);

    LOG_ERR(nullptr,"ModelSine::AnalysisAndSynthesis Synthesis done");
    SampleListType result(y.begin(),y.end());

    LOG_ERR(nullptr,"ModelSine::AnalysisAndSynthesis Synthesis result size = %lld",result.size());

    return result;
  }

}

File: src/python/ModelSine.py
import pdb
import numpy as np 
import matplotlib.pyplot as plt
from ispeech.source import *
from ispeech.transform import *

def peakDetection(mX,t):
    size = len(mX)
    ploc = np.zeros(size,dtype=int)
    j = 0;
    for i in range(1,size-1):
        if mX[i] > t:
            if mX[i]>mX[i-1] and mX[i] > mX[i+1]:
                ploc[j] = int(i)
                j+=1

    return ploc;
def peakDetection2(mX, t):
    thresh = np.where(np.greater(mX[1:-1],t), mX[1:-1], 0) 
    next_minor = np.where(mX[1:-1]>mX[2:], mX[1:-1], 0)   
    prev_minor = np.where(mX[1:-1]>mX[:-2], mX[1:-1], 0)
    ploc = thresh * next_minor * prev_minor
    ploc = ploc.nonzero()[0] + 1
    return ploc

#parabola
def peakInterp(mX, pX, ploc):
    val = mX[ploc]
    lval = mX[ploc-1]
    rval = mX[ploc+1]
    iploc = ploc + 0.5*(lval-rval)/(lval-2*val+rval)
    ipmag = val - 0.25*(lval-rval)*(iploc-ploc)
    ipphase = np.interp(iploc, np.arange(0, pX.size), pX)
    return iploc, ipmag, ipphase

#this function same to np.interp
def Interp(x,xp,fp):
    ix = np.zeros(len(x))
    for i in range(len(x)):
        if x[i] <= xp[0]:
            ix[i]= fp[0]
            continue
        if x[i] >= xp[-1]:
            ix[i]=fp[-1]         
        for j in range(len(xp)-1):
            if x[i] > xp[j] and x[i] < xp[j+1]:
                ix[i]= fp[j]+(x[i]-xp[j])/(xp[j+1]-xp[j])*(fp[j+1]-fp[j])
                continue
            if x[i] == xp[j]:
                ix[i]=fp[j]
                continue
    return ix


def sineModel(x, fs, w, N, t):
    # half analysis window size by rounding
    hM1 = int(math.floor((w.size+1)/2))
    # half analysis window size by floor
    hM2 = int(math.floor(w.size/2))
    # FFT size for synthesis (even)    
    Ns = 512

    # Hop size used for analysis and synthesis
    H = Ns//4
    print('H='+str(H))
    # half of synthesis FFT size
    hNs = Ns//2
    print('hNs = ' + str(hNs))

    # init sound pointer in middle of anal window
    pin = max(hNs, hM1)
    # last sample to start a frame       
    pend = x.size - max(hNs, hM1)
    # initialize buffer for FFT
    fftbuffer = np.zeros(N)
    # initialize output sound frame
    yw = np.zeros(Ns)

    # initialize output array
    y = np.zeros(x.size)
    # normalize analysis window
    w = w / sum(w)
        
    # initialize synthesis window
    sw = np.zeros(Ns)

    # triangular window
    ow = triang(2*H)

    # add triangular window
    sw[hNs-H:hNs+H] = ow

    # blackmanharris window
    bh = blackmanharris(Ns)
    # normalized blackmanharris window
    bh = bh / sum(bh)

    # normalized synthesis window
    sw[hNs-H:hNs+H] = sw[hNs-H:hNs+H] / bh[hNs-H:hNs+H]
    # while input sound pointer is within sound
    while pin<pend:             
        #-----analysis-----
        # select frame
        x1 = x[pin-hM1:pin+hM2]
        # compute dft
        mX, pX = self.dftAnal(x1, w, N)
        # detect locations of peaks
        ploc = peakDetection(mX, t)

        iploc, ipmag, ipphase = peakInterp(mX, pX, ploc)
        # convert peak locations to Hertz
        ipfreq = fs*iploc/float(N)
        #-----synthesis-----
        # generate sines in the spectrum
        Y = self.genSpecSines(ipfreq, ipmag, ipphase, Ns, fs)
        print('Y size = ' + str(Y.size))
        print(np.real(Y))
        # compute inverse FFT
        fftbuffer = np.real(ifft(Y))
        # undo zero-phase window
        yw[:hNs-1] = fftbuffer[hNs+1:]            
        yw[hNs-1:] = fftbuffer[:hNs+1]
        # overlap-add and apply a synthesis window
        #y[pin-hNs:pin+hNs] += sw*yw
        y[pin-hNs:pin+hNs] += sw*yw


        # advance sound pointer
        pin += H

    return y

xp=[1,2,3]
fp=[3,2,0]
x=[0,1,1.57,2.5,3.14]
ix=np.interp(x,xp,fp)
ix2=Interp(x,xp,fp)
print(ix)
print(ix2)

                
def test_peakInterp():
  
    fs = 44100
    f = 500
    N = 256;
    sine = SineGenerator(fs)
    sine.setFrequency(f)
    sine.setAmplitude(1.0)
    sine.setPhase(1.0)
    sine.generate(N)

    hamming = HammingWindow(N)

    sine = np.array(sine)
    hamming = np.array(hamming)

    sine = sine*hamming
    
    dft = Dft(N)
   
    spectrum_sine=dft.fft(sine)

    spectrum_sine = np.array(spectrum_sine);

    spectrum_sine = abs(spectrum_sine)
    sine_idft = dft.ifft(spectrum_sine)

    plt.figure(1)
    plt.subplot(3,1,1)
    plt.plot(np.arange(N),sine)
    ploc = peakDetection2(sine,-100)
    
    plt.plot(ploc, sine[ploc], marker='x', color='b', linestyle='', markeredgewidth=1.5)   
    plt.title('sine signal and peaks')

    plt.subplot(3,1,2)
    plt.plot(np.arange(spectrum_sine.size)*fs/N,abs(spectrum_sine),'g',label='spectrum_sine')
    ploc = peakDetection2(abs(spectrum_sine),-100)
    plt.plot(ploc*fs/N, abs(spectrum_sine[ploc]), marker='x', color='b', linestyle='', markeredgewidth=1.5)   

    plt.title('DFT and peaks')

    plt.subplot(3,1,3)
    plt.plot(np.arange(spectrum_sine.size)*fs/N,abs(spectrum_sine),'g',label='spectrum_sine')

    iploc,ipmag,ipphase = peakInterp(spectrum_sine,spectrum_sine,ploc)
    iploc = np.array(iploc,dtype=int)
    plt.plot(iploc*fs/N, abs(ipphase), marker='x', color='b', linestyle='', markeredgewidth=1.5)
    print(iploc.size)

    print(ipphase.size)
    

    plt.legend()
    plt.show()

    
def test_peakDetection():
  
    fs = 44100
    f = 500
    N = 256;
    sine = SineGenerator(fs)
    sine.setFrequency(f)
    sine.setAmplitude(1.0)
    sine.setPhase(1.0)
    sine.generate(N)

    hamming = HammingWindow(N)

    sine = np.array(sine)
    hamming = np.array(hamming)

    sine = sine*hamming
    
    dft = Dft(N)
   
    spectrum_sine=dft.fft(sine)

    spectrum_sine = np.array(spectrum_sine);

   
    sine_idft = dft.ifft(spectrum_sine)

    plt.figure(1)
    plt.subplot(2,1,1)
    plt.plot(np.arange(N),sine)
    ploc = peakDetection2(sine,-100)
    
    plt.plot(ploc, sine[ploc], marker='x', color='b', linestyle='', markeredgewidth=1.5)   
    plt.title('sine signal')

    plt.subplot(2,1,2)
    plt.plot(np.arange(spectrum_sine.size)*fs/N,abs(spectrum_sine),'g',label='spectrum_sine')
    ploc = peakDetection2(abs(spectrum_sine),-100)
    plt.plot(ploc*fs/N, abs(spectrum_sine[ploc]), marker='x', color='b', linestyle='', markeredgewidth=1.5)   

    plt.title('DFT')

    plt.legend()
    plt.show()

#test_peakDetection()
#test_peakInterp()
